src/Features/Core/Portable/ExtractMethod/SelectionResult.cs

```diff
@@ -15,213 +13,203 @@
+        public SemanticDocument SemanticDocument { get; private set; } = document;
+        public TextSpan OriginalSpan { get; } = originalSpan;
+        public TextSpan FinalSpan { get; } = finalSpan;
+        public SelectionType SelectionType { get; } = selectionType;
+        public bool SelectionChanged { get; } = selectionChanged;
 
-    public abstract bool ContainingScopeHasAsyncKeyword();
+        protected abstract ISyntaxFacts SyntaxFacts { get; }
+        protected abstract bool UnderAnonymousOrLocalMethod(SyntaxToken token, SyntaxToken firstToken, SyntaxToken lastToken);
 
-    public abstract SyntaxNode GetContainingScope();
-    public abstract SyntaxNode GetOutermostCallSiteContainerToProcess(CancellationToken cancellationToken);
+        public abstract TStatementSyntax GetFirstStatementUnderContainer();
+        public abstract TStatementSyntax GetLastStatementUnderContainer();
 
-    public abstract (ITypeSymbol returnType, bool returnsByRef) GetReturnType();
+        public abstract bool ContainingScopeHasAsyncKeyword();
 
-    public ITypeSymbol GetContainingScopeType()
-    {
-        var (typeSymbol, _) = GetReturnType();
-        return typeSymbol;
-    }
+        public abstract SyntaxNode GetContainingScope();
+        public abstract SyntaxNode GetOutermostCallSiteContainerToProcess(CancellationToken cancellationToken);
 
-    public virtual SyntaxNode GetNodeForDataFlowAnalysis() => GetContainingScope();
+        public abstract (ITypeSymbol? returnType, bool returnsByRef) GetReturnType();
 
-    public SelectionResult<TStatementSyntax> With(SemanticDocument document)
-    {
-        if (SemanticDocument == document)
+        public ITypeSymbol? GetContainingScopeType()
         {
-            return this;
+            var (typeSymbol, _) = GetReturnType();
+            return typeSymbol;
         }
 
-        var clone = (SelectionResult<TStatementSyntax>)MemberwiseClone();
-        clone.SemanticDocument = document;
+        public bool IsExtractMethodOnExpression => this.SelectionType == SelectionType.Expression;
+        public bool IsExtractMethodOnSingleStatement => this.SelectionType == SelectionType.SingleStatement;
+        public bool IsExtractMethodOnMultipleStatements => this.SelectionType == SelectionType.MultipleStatements;
 
-        return clone;
-    }
+        public virtual SyntaxNode? GetNodeForDataFlowAnalysis() => GetContainingScope();
 
-    public SyntaxToken GetFirstTokenInSelection()
-        => SemanticDocument.GetTokenWithAnnotation(FirstTokenAnnotation);
+        public SelectionResult With(SemanticDocument document)
+        {
+            if (SemanticDocument == document)
+            {
+                return this;
+            }
 
-    public SyntaxToken GetLastTokenInSelection()
-        => SemanticDocument.GetTokenWithAnnotation(LastTokenAnnotation);
+            var clone = (SelectionResult)MemberwiseClone();
+            clone.SemanticDocument = document;
```

Comment:
i hate this method.  it's not ok that we're doing a MemberwiseClone. t his will be fixed up later.